### ISSUE TITLE
R crash when importing pymer4 module

### ISSUE THREAD

**Issue by dramanica** [rstudio/reticulate#1236]
R completely crashes when trying to import the pymer4 module. I have tried this both in R 4.2.1 (direct install on Ubuntu 20.04) as well as R 4.1.3 from within a conda env (same operating system), with all packages up to date. Here is a quick reproducible example:

```
library(reticulate)
# use mamba to quickly solve the environment, conda is very slow to resolve for pymer4
options(reticulate.conda_binary = Sys.which("mamba"))
conda_create("pymer_reticulate", python_version="3.9")
conda_install(envname="pymer_reticulate", packages="pymer4",channel=c("ejolly","conda-forge","defaults"))
```
If a load the environment, I get:
```
use_condaenv("pymer_reticulate")

/tmp/RtmpLi3ak3/file129314983774f.sh: 5: /home/andrea/miniconda3/envs/pymer_reticulate/etc/conda/activate.d/activate-r-base.sh: [[: not found
```
I have tested the environment running various python scripts, including using `rpy2 `on which `pymer4` is based , and everything works just fine despite the warning when I load the environment about the missing `activate-r-base.sh`. (not sure where that comes from).
However, importing anything from `pymer4` gives a crash:

```
 py_run_string("from pymer4.test_install import test_install")

/home/andrea/miniconda3/envs/pymer_reticulate/lib/python3.9/site-packages/rpy2/rinterface_lib/embedded.py:134: UserWarning: R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
  warnings.warn(
[73997:73997:20220712,055746.587536:ERROR process_memory_range.cc:86] read out of range
[73997:73997:20220712,055746.587573:ERROR elf_image_reader.cc:558] missing nul-terminator
[73997:73997:20220712,055746.587678:ERROR elf_dynamic_array_reader.h:61] tag not found
[73997:73997:20220712,055746.590287:ERROR elf_dynamic_array_reader.h:61] tag not found
[73997:73997:20220712,055746.590335:ERROR elf_dynamic_array_reader.h:61] tag not found
[73997:73997:20220712,055746.590374:ERROR elf_dynamic_array_reader.h:61] tag not found
... many more instances of the same elf_dynamic_array_reader.h error
[73997:73997:20220712,055746.598358:ERROR elf_dynamic_array_reader.h:61] tag not found
[73997:73998:20220712,055746.625116:ERROR directory_reader_posix.cc:42] opendir: No such file or directory (2)
```
And R goes down hard.

In bash, I can run the commands directly in python without any problem, so the conda environment is fully functional:
```
(base) andrea@laptop:~$ conda activate pymer_reticulate
(pymer_reticulate) andrea@laptop:~$ python -c "from pymer4.test_install import test_install; test_install()"

Pymer4 installation working successfully!
```
Any help would be much appreciated. Thanks!

**Comment by t-kalinowski** [comment]
Hi @dramanica, Thanks for reporting.

I just tried with vanilla `conda` and pymer4 installed successfully for me. However, then activating the environment gave a different error:
```r
> library(reticulate)
> use_condaenv("pymer_reticulate")
/tmp/RtmpaWyNaZ/file29026d9a7db8.sh: 5: /home/tomasz/.local/share/r-miniconda/envs/pymer_reticulate/etc/conda/activate.d/activate-r-base.sh: [[: not found
```

Which I can also reproduce outside of R:
```bash
$ sh /home/tomasz/.local/share/r-miniconda/envs/pymer_reticulate/etc/conda/activate.d/activate-r-base.sh
/home/tomasz/.local/share/r-miniconda/envs/pymer_reticulate/etc/conda/activate.d/activate-r-base.sh: 5: [[: not found
```

The contents of: [`activate-r-base.sh`](https://github.com/conda-forge/r-base-feedstock/blob/main/recipe/activate-r-base.sh):
```bash
#!/usr/bin/env sh
...
```
Turns out, despite specifying `sh`, it's not really a portable shell script because it requires `[[`. Regardless, the error is unwittingly helpful in this case because if you're not using builds of R and Rstudio provided by conda, then we recommend letting reticulate activate the conda environment, rather than attempting to launch a non-conda-build of R from an already activated conda environment.

`mamba` compatibility is out-of-scope for what we can officially support, but I'm happy to accept PRs if the changes required to add compatibility are minor.

Note, after working around the `[[` issue, there is still the separate issue where `pymer4` segfaults, in a way that's reminiscent of rpy2 process sharing issue that was fixed recent reticulate and rpy2 releases. After looking into it a bit however, it appears the fix will need to be in pymer4 sources (or the way it's packaged in the custom conda channel), and not in reticulate or rpy2. Note that rpy2 installed via pip, without conda in the mix, works with reticulate without any issues. Attempting to install pymer4 via pip was unsuccessful for me.

**Comment by dramanica** [comment]
@t-kalinowski Thanks for the prompt reply. Would you be able to provide a steer in what I should look for in the pymer4 sources? What kind of calls could lead to such process sharing issues? Even a reference for any old issue about the rpy2 problems that I could use for inspiration would be helpful. Many thanks.

**Comment by dramanica** [comment]
@t-kalinowski Thanks for the tips! So, I did what you suggested; unfortunately, I still manage to segfault.
I have a local installation of R on linux (v 4.2.1), with the required R libraries from CRAN:
`install.packages(c("lmerTest","emmeans"))`
and the latest version of `reticulate`

In bash, I use a conda env to localise python, but install all packages with pip:
```
conda create -n pymer4_reticulate python=3.9
conda activate pymer4_reticulate
pip install rpy2
pip install git+https://github.com/ejolly/pymer4.git@dev_rpy2_3.5.1
```
This gives us an env with `rpy` v3.5.2 (ignore the pymer4 git tag mentioning 3.5.1), and I can successfully use `pymer4`:
`python -c "from pymer4.test_install import test_install; test_install()"`

There is no R in the conda env, so any invocation of R goes to the default Linux installation.
However, if I try to go via `R`:
```
R
library(reticulate)
use_condaenv("pymer4_reticulate")
py_run_string("from pymer4.test_install import test_install")

```
I get a segfault. Any suggestion on where to poke next? Many thanks for your help.

**Comment by t-kalinowski** [comment]
You can try to run R with gdb (or lldb on Mac) to see where the segfault is coming from. This is a nice getting-started guide to debugging C code with R and lldb: https://kevinushey.github.io/blog/2015/04/13/debugging-with-lldb/

**Comment by t-kalinowski** [comment]
One followup: A recent source of segfaults has been NumPy incompatibility with R's BLAS. A workaround is to install numpy from source:

```r
reticulate::py_install(
  "numpy",
  envname = envname,
  pip = TRUE,
  pip_options = c("--force-reinstall", "--no-binary numpy")
)

```

**Comment by dramanica** [comment]
@t-kalinowski Thanks for the suggestion. I just tried reinstalling numpy from source, but that does not seem to fix the problem. I tried to have a look at the backtrace from gdb:
[New Thread 0x7fffd4937700 (LWP 30602)]
warning: Loadable section ".note.gnu.property" outside of ELF segments

```
Thread 1 "R" received signal SIGSEGV, Segmentation fault.
_PyTrash_begin (tstate=tstate@entry=0x0, op=op@entry=0x7fffd6ad7d60)
    at /usr/local/src/conda/python-3.9.13/Objects/object.c:2125
2125	/usr/local/src/conda/python-3.9.13/Objects/object.c: No such file or directory.
(gdb) bt
#0  _PyTrash_begin (tstate=tstate@entry=0x0, op=op@entry=0x7fffd6ad7d60)
    at /usr/local/src/conda/python-3.9.13/Objects/object.c:2125
#1  0x00007fffda9140bf in meth_dealloc (m=0x7fffd6ad7d60)
    at /usr/local/src/conda/python-3.9.13/Objects/methodobject.c:166
#2  0x00007ffff7c84e31 in R_RunWeakRefFinalizer () from /usr/lib/R/lib/libR.so
#3  0x00007ffff7c850ac in ?? () from /usr/lib/R/lib/libR.so
#4  0x00007ffff7c39818 in ?? () from /usr/lib/R/lib/libR.so
#5  0x00007ffff7c43b98 in Rf_eval () from /usr/lib/R/lib/libR.so
#6  0x00007ffff7c45a4f in ?? () from /usr/lib/R/lib/libR.so
#7  0x00007ffff7c46942 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#8  0x00007ffff7c313ce in ?? () from /usr/lib/R/lib/libR.so
#9  0x00007ffff7c43b98 in Rf_eval () from /usr/lib/R/lib/libR.so
#10 0x00007ffff7c45a4f in ?? () from /usr/lib/R/lib/libR.so
#11 0x00007ffff7c46942 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#12 0x00007ffff7c313ce in ?? () from /usr/lib/R/lib/libR.so
#13 0x00007ffff7c43b98 in Rf_eval () from /usr/lib/R/lib/libR.so
#14 0x00007ffff7c45a4f in ?? () from /usr/lib/R/lib/libR.so
#15 0x00007ffff7c46942 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#16 0x00007ffff7c313ce in ?? () from /usr/lib/R/lib/libR.so
#17 0x00007ffff7c43b98 in Rf_eval () from /usr/lib/R/lib/libR.so
#18 0x00007ffff7c45a4f in ?? () from /usr/lib/R/lib/libR.so
#19 0x00007ffff7c46942 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#20 0x00007ffff7c43dbf in Rf_eval () from /usr/lib/R/lib/libR.so
#21 0x00007ffff7bd31a2 in ?? () from /usr/lib/R/lib/libR.so
#22 0x00007ffff7bd4a4a in R_ToplevelExec () from /usr/lib/R/lib/libR.so
#23 0x00007ffff7bd4abd in R_tryEval () from /usr/lib/R/lib/libR.so
#24 0x00007fffd61cb1bf in _cffi_f_R_tryEval ()
   from /home/andrea/miniconda3/envs/pymer4_reticulate/lib/python3.9/site-packages/_rinterface_cffi_api.abi3.so
#25 0x00007fffda93107a in cfunction_call (func=0x7ffff39b7d10, 
    args=<optimised out>, kwargs=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Objects/methodobject.c:552
#26 0x00007fffda91fde6 in _PyObject_MakeTpCall (tstate=0x55555ecdaf40, 
    callable=0x7ffff39b7d10, args=<optimised out>, nargs=<optimised out>, 
    keywords=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Include/object.h:430
#27 0x00007fffda91cdf5 in _PyObject_VectorcallTstate (kwnames=0x0, 
    nargsf=9223372036854775811, args=0x55555f81fc90, callable=0x7ffff39b7d10, 
    tstate=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Include/cpython/abstract.h:116
#28 _PyObject_VectorcallTstate (kwnames=0x0, nargsf=9223372036854775811, 
    args=0x55555f81fc90, callable=0x7ffff39b7d10, tstate=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Include/cpython/abstract.h:103
#29 PyObject_Vectorcall (kwnames=0x0, nargsf=9223372036854775811, 
    args=0x55555f81fc90, callable=0x7ffff39b7d10)
    at /usr/local/src/conda/python-3.9.13/Include/cpython/abstract.h:127
#30 call_function (kwnames=0x0, oparg=<optimised out>, 
    pp_stack=<synthetic pointer>, tstate=0x55555ecdaf40)
    at /usr/local/src/conda/python-3.9.13/Python/ceval.c:5077
#31 _PyEval_EvalFrameDefault (tstate=<optimised out>, f=<optimised out>, 
    throwflag=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Python/ceval.c:3489
#32 0x00007fffda916f69 in _PyEval_EvalFrame (throwflag=0, f=0x55555f81fac0, 
    tstate=0x55555ecdaf40)
    at /usr/local/src/conda/python-3.9.13/Include/internal/pycore_ceval.h:40
#33 _PyEval_EvalCode (tstate=<optimised out>, _co=<optimised out>, 
    globals=<optimised out>, locals=<optimised out>, args=<optimised out>, 
    argcount=<optimised out>, kwnames=<optimised out>, kwargs=<optimised out>, 
    kwcount=<optimised out>, kwstep=<optimised out>, defs=<optimised out>, 
    defcount=<optimised out>, kwdefs=<optimised out>, closure=<optimised out>, 
    name=<optimised out>, qualname=<optimised out>)
    at /usr/local/src/conda/python-3.9.13/Python/ceval.c:4329
--Type <RET> for more, q to quit, c to continue without paging--
```
I am taking it that from `#1 `to `#23` we are looking at R bailing out, so the first interesting line is `#24,` which where python is reporting back to R via its foreign interface (I would presume that is from rpy2). Everything before that is core python (until you get to the very beginning, where python was initiated by `py_run_string_impl`. So, if I was debugging one of my small C programs, I would be interested in knowing what went from `#24` to `#25`, but I have to confess that I am a bit of a loss on how to do that within the current set up.

**Comment by t-kalinowski** [comment]
Hmm, looking at the backtrace, my best guess is that the segfault is being triggered by Python's garbage collector attempting to finalize a Python object that has already been deallocated. (triggered likely from `R_RunWeakRefFinalizer()`). 

This thread looks related: https://github.com/python/cpython/issues/86181

This warrents some investigation. It would be very helpful if you could make a small reproducible example so I can reproduce locally without needing to install the full `pymer4` module.

**Comment by dramanica** [comment]
@t-kalinowski Just to be clear, what you would need ideally is a small python script that replicates what `pymer4` does with `rpy2` and gives a similar crash, correct? It might require a bit of digging into the `pymer4` code. 
Happy to have a go at it, but just wanted to make sure I know what i am trying to achieve before wasting time.

**Comment by t-kalinowski** [comment]
Yes, a small python script and/or minimal python module would be ideal. One strategy would be to start with the pymer4 module, and start deleting parts of it until you've isolated the segfault.

**Comment by t-kalinowski** [comment]
Hmm, I can't reproduce the segfault with a pyenv installed python `reticulate::install_python()`. 

```
tomasz@horse:~
$ Rscript -e 'library(reticulate); py_run_file("test_seq.py"); py_config()'
/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/rpy2/rinterface_lib/embedded.py:260: UserWarning: R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
  warnings.warn(msg)
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
blah
python:         /home/tomasz/.virtualenvs/r-reticulate/bin/python
libpython:      /home/tomasz/.pyenv/versions/3.9.13/lib/libpython3.9.so
pythonhome:     /home/tomasz/.virtualenvs/r-reticulate:/home/tomasz/.virtualenvs/r-reticulate
version:        3.9.13 (main, Jul 12 2022, 17:17:48)  [GCC 9.4.0]
numpy:          /home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/numpy
numpy_version:  1.22.4

python versions found: 
 /home/tomasz/.virtualenvs/r-reticulate/bin/python
 /usr/bin/python3
 /usr/bin/python
 /home/tomasz/.virtualenvs/r-tensorflow-site/bin/python
tomasz@horse:~
$ cat test_seq.py 
from rpy2.robjects.packages import importr
base = importr("base")
base.cat("blah\n")
```

**Comment by dramanica** [comment]
Hmm, something strange is going on. So, I tried installing python via `reticulate::install_python().` and everything worked. However, if I issue the command  `py_run_file("test_seq.py")` a second time, I get a segfault. After that, I get a segfault immediately on the first run of `py_run_file`. I tested by creating a brand new virtualenv, and I can replicate the sequence of running fine on first instance, and then crashing afterwards.

**Comment by t-kalinowski** [comment]
@lgautier Can you please take a look? It looks like rpy2 attempts to initialize the R session more than once:

minimal reprex:
```bash
Rscript -e 'for (i in 1:100) reticulate::py_run_string("from rpy2.robjects.packages import importr; importr(\"base\").cat(\"Hello!\")")'
```
Output I see:
```
$ Rscript -e 'for (i in 1:100) reticulate::py_run_string("from rpy2.robjects.packages import importr; importr(\"base\").cat(\"Hello!\")")'
/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/rpy2/rinterface_lib/embedded.py:260: UserWarning: R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
  warnings.warn(msg)
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
Hello!
 *** caught segfault ***
address 0xb8, cause 'memory not mapped'
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)
```

**Comment by lgautier** [comment]
@t-kalinowski 

> ```
> R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
> ```

This error message should mean that `rpy2` identifies that libR loaded in the current process was initialized elsewhere. This will be the case with `reticulate` as libR is initialized when calling R. It will then proceed without trying to initialize it a second time:
https://github.com/rpy2/rpy2/blob/master/rpy2/rinterface_lib/embedded.py#L258

This a bit of guess though. It relies on R_NilValue not being NULL as there is no formal way to check this in the R API. This is also unable to detect if the embedded R was also ended. Trying to use libR for pretty much anything after R was ended is likely to quickly end with a segfault.

The other likely causes are:
- conda and/or conda packaging (e.g., wrong libR picked at runtime)
- concurrent access to libR (multithreading is a bit no-no, but multiple processes could also get you in trouble).

**Comment by t-kalinowski** [comment]
Any thoughts about why the warning is issued twice? Would you expect https://github.com/rpy2/rpy2/blob/cf656a0c6fe420b96204adf29a065f3e0428d04e/rpy2/rinterface_lib/embedded.py#L252 to return `True` one the 2nd time though this code path?

**Comment by lgautier** [comment]
There is a call to `warning.warn()` and to `logger.warn()`:
https://github.com/rpy2/rpy2/blob/master/rpy2/rinterface_lib/embedded.py#L260

`logger.warn()` should be enough and I should fix that.

**Comment by lgautier** [comment]
> ```bash
> Rscript -e 'for (i in 1:100) reticulate::py_run_string("from rpy2.robjects.packages import importr; importr(\"base\").cat(\"Hello!\")")'
> ```

Is the loop needed to trigger the issue? If the case, my guess would be that circular references straddling the two languages create trouble at garbage collection time.

**Comment by lgautier** [comment]
> ```bash
> Rscript -e 'for (i in 1:100) reticulate::py_run_string("from rpy2.robjects.packages import importr; importr(\"base\").cat(\"Hello!\")")'
> ```

Is the loop needed to trigger the issue? If the case, my guess would be that circular references straddling the two languages create trouble at garbage collection time.

**Comment by t-kalinowski** [comment]
Yes, the loop is needed, but only for one iteration. The segfault happens on the 2nd import of rpy2.

**Comment by lgautier** [comment]
Things are working beyond 2 iterations on my end:

```
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!^CError in py_run_string_impl(code, local, convert) : KeyboardInterrupt
Calls: <Anonymous> -> py_run_string_impl
In addition: There were 50 or more warnings (use warnings() to see the first 50)
Execution halted
```

**Comment by lgautier** [comment]
My R is a bit dated: `R version 4.1.1 Patched (2021-09-11 r80887)`.
Python 3.8.10.

**Comment by t-kalinowski** [comment]
Interesting. Trying it out with R 4.1.1 I get through a few more iterations before the default, but it still segfaults.


```
$ /opt/R/4.1.1/bin/Rscript - << EOF
> sessionInfo()
> reticulate::py_config()
> for(i in 1:10)
> reticulate::py_run_string("from rpy2.robjects.packages import importr; importr('base').cat('Hello!')")
> EOF
R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libmkl_rt.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.1.1
python:         /home/tomasz/.virtualenvs/r-tensorflow-site/bin/python
libpython:      /home/tomasz/.pyenv/versions/3.9.13/lib/libpython3.9.so
pythonhome:     /home/tomasz/.virtualenvs/r-tensorflow-site:/home/tomasz/.virtualenvs/r-tensorflow-site
version:        3.9.13 (main, Jul 12 2022, 17:17:48)  [GCC 9.4.0]
numpy:          /home/tomasz/.virtualenvs/r-tensorflow-site/lib/python3.9/site-packages/numpy
numpy_version:  1.23.1

NOTE: Python version was forced by use_python function
/home/tomasz/.virtualenvs/r-tensorflow-site/lib/python3.9/site-packages/rpy2/rinterface_lib/embedded.py:260: UserWarning: R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
  warnings.warn(msg)
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
Hello!Hello!Hello!
 *** caught segfault ***
address 0xb8, cause 'memory not mapped'
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)
```

**Comment by lgautier** [comment]
I seem to many more iterations:

```
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!Hello!^CError in py_run_string_impl(code, local, convert) : KeyboardInterrupt
```


reticulate is 1.25-9000
rpy2 is the dev branch.

All this running on Linux.

**Comment by t-kalinowski** [comment]
To make sure we're on the same page, I installed R like this:
```
export R_VERSION=4.1.1 # you can also use 4.2.1
curl -O https://cdn.rstudio.com/r/ubuntu-2004/pkgs/r-${R_VERSION}_1_amd64.deb
sudo gdebi r-${R_VERSION}_1_amd64.deb
```
To install reticulate, from the R shell:
```R
install.packages("remotes")
remotes::install_github("rstudio/reticulate")
```
You can point reticulate at whatever python you are using with `RETICULATE_PYTHON`

e.g
```bash
export RETICULATE_PYTHON=/usr/bin/python3
Rscript -e 'reticulate::py_config()'
```



```

**Comment by kevinushey** [comment]
In case it helps:

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xb8)
  * frame #0: 0x00000001128ae16c Python`meth_dealloc + 100
    frame #1: 0x0000000105218500 libR.dylib`R_RunWeakRefFinalizer(w=<unavailable>) at memory.c:1500:2 [opt]
    frame #2: 0x0000000105218a40 libR.dylib`RunFinalizers at memory.c:1567:3 [opt]
    frame #3: 0x0000000105218b10 libR.dylib`R_RunPendingFinalizers at memory.c:1603:2 [opt] [artificial]
    frame #4: 0x00000001051c36ec libR.dylib`bcEval [inlined] bc_check_sigint at eval.c:5550:5 [opt]
    frame #5: 0x00000001051c36e4 libR.dylib`bcEval(body=0x0000000124dd4ac0, rho=<unavailable>, useCache=<unavailable>) at eval.c:6744:4 [opt]
    frame #6: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x0000000124dd4ac0, rho=0x00000001107d32a0) at eval.c:748:8 [opt]
    frame #7: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x00000001234ef2b8, newrho=0x00000001107d32a0, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #8: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x00000001234ef2b8, op=0x0000000124dd4c80, arglist=0x00000001107d3498, rho=0x00000001107c0058, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #9: 0x00000001051c7770 libR.dylib`bcEval(body=0x00000001234d7288, rho=<unavailable>, useCache=<unavailable>) at eval.c:7104:12 [opt]
    frame #10: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x00000001234d7288, rho=0x00000001107c0058) at eval.c:748:8 [opt]
    frame #11: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x00000001234f9510, newrho=0x00000001107c0058, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #12: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x00000001234f9510, op=0x00000001234d7410, arglist=0x00000001107c0218, rho=0x00000001107c0330, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #13: 0x00000001051c7770 libR.dylib`bcEval(body=0x00000001234f9c10, rho=<unavailable>, useCache=<unavailable>) at eval.c:7104:12 [opt]
    frame #14: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x00000001234f9c10, rho=0x00000001107c0330) at eval.c:748:8 [opt]
    frame #15: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x0000000123492720, newrho=0x00000001107c0330, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #16: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x0000000123492720, op=0x00000001234f9d28, arglist=0x00000001107c05d0, rho=0x00000001105cf270, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #17: 0x00000001051c7770 libR.dylib`bcEval(body=0x0000000123460518, rho=<unavailable>, useCache=<unavailable>) at eval.c:7104:12 [opt]
    frame #18: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x0000000123460518, rho=0x00000001105cf270) at eval.c:748:8 [opt]
    frame #19: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x00000001105cf628, newrho=0x00000001105cf270, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #20: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x00000001105cf628, op=0x0000000123460978, arglist=0x00000001105cf580, rho=0x00000001230468c8, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #21: 0x00000001051c085c libR.dylib`Rf_eval(e=0x00000001105cf628, rho=0x00000001230468c8) at eval.c:871:12 [opt]
    frame #22: 0x000000010517ceb0 libR.dylib`protectedEval(d=0x000000016b4d21a8) at context.c:831:17 [opt]
    frame #23: 0x000000010517cd2c libR.dylib`R_ToplevelExec(fun=(libR.dylib`protectedEval at context.c:825), data=0x000000016b4d21a8) at context.c:779:2 [opt]
    frame #24: 0x000000010517ce2c libR.dylib`R_tryEval(e=<unavailable>, env=<unavailable>, ErrorOccurred=0x0000600002fd4298) at context.c:845:10 [opt]
    frame #25: 0x000000012215d634 _rinterface_cffi_api.abi3.so`_cffi_f_R_tryEval + 676
    frame #26: 0x00000001128ae428 Python`cfunction_call + 96
    frame #27: 0x000000011285f320 Python`_PyObject_MakeTpCall + 132
    frame #28: 0x0000000112951010 Python`_PyEval_EvalFrameDefault + 28164
    frame #29: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #30: 0x000000011285ff08 Python`_PyFunction_Vectorcall + 348
    frame #31: 0x000000011285fad0 Python`PyVectorcall_Call + 184
    frame #32: 0x000000011294accc Python`_PyEval_EvalFrameDefault + 2752
    frame #33: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #34: 0x000000011285ff08 Python`_PyFunction_Vectorcall + 348
    frame #35: 0x000000011285f610 Python`_PyObject_FastCallDictTstate + 208
    frame #36: 0x00000001128d1ca8 Python`slot_tp_call + 188
    frame #37: 0x000000011285fca8 Python`_PyObject_Call + 128
    frame #38: 0x000000011294accc Python`_PyEval_EvalFrameDefault + 2752
    frame #39: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #40: 0x000000011285ff08 Python`_PyFunction_Vectorcall + 348
    frame #41: 0x000000011285f5a0 Python`_PyObject_FastCallDictTstate + 96
    frame #42: 0x00000001128d32dc Python`slot_tp_init + 188
    frame #43: 0x00000001128d8eb8 Python`type_call + 300
    frame #44: 0x000000011285f320 Python`_PyObject_MakeTpCall + 132
    frame #45: 0x000000011294fffc Python`_PyEval_EvalFrameDefault + 24048
    frame #46: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #47: 0x000000011285ff08 Python`_PyFunction_Vectorcall + 348
    frame #48: 0x000000011294eac0 Python`_PyEval_EvalFrameDefault + 18612
    frame #49: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #50: 0x000000011285ff08 Python`_PyFunction_Vectorcall + 348
    frame #51: 0x000000011294ea98 Python`_PyEval_EvalFrameDefault + 18572
    frame #52: 0x000000011294920c Python`_PyEval_EvalCode + 452
    frame #53: 0x00000001129a0ca0 Python`run_eval_code_obj + 136
    frame #54: 0x00000001129a0bd0 Python`run_mod + 112
    frame #55: 0x00000001129a1960 Python`PyRun_StringFlags + 140
    frame #56: 0x00000001076ccd9c reticulate.so`py_run_string_impl(code="from rpy2.robjects.packages import importr; importr('base').cat('Hello!')", local=false, convert=true) at python.cpp:2718:21 [opt]
    frame #57: 0x00000001076aca20 reticulate.so`::_reticulate_py_run_string_impl(codeSEXP=<unavailable>, localSEXP=<unavailable>, convertSEXP=0x00000001203ed2a8) at RcppExports.cpp:544:34 [opt]
    frame #58: 0x0000000105191bf4 libR.dylib`R_doDotCall(ofun=<unavailable>, nargs=3, cargs=0x000000016b4d4130, call=0x0000000116d90898) at dotcode.c:607:17 [opt]
    frame #59: 0x00000001051da2dc libR.dylib`bcEval(body=0x0000000116d90940, rho=<unavailable>, useCache=<unavailable>) at eval.c:7692:21 [opt]
    frame #60: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x0000000116d90940, rho=0x000000012038b6a8) at eval.c:748:8 [opt]
    frame #61: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x0000000116b6db18, newrho=0x000000012038b6a8, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #62: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x0000000116b6db18, op=0x0000000116d90d30, arglist=0x000000012038b868, rho=0x000000012038bb40, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #63: 0x00000001051c7770 libR.dylib`bcEval(body=0x0000000116b6ddb8, rho=<unavailable>, useCache=<unavailable>) at eval.c:7104:12 [opt]
    frame #64: 0x00000001051c0588 libR.dylib`Rf_eval(e=0x0000000116b6ddb8, rho=0x000000012038bb40) at eval.c:748:8 [opt]
    frame #65: 0x00000001051ddb94 libR.dylib`R_execClosure(call=0x00000001166f93b8, newrho=0x000000012038bb40, sysparent=<unavailable>, rho=<unavailable>, arglist=<unavailable>, op=<unavailable>) at eval.c:0 [opt]
    frame #66: 0x00000001051dc3cc libR.dylib`Rf_applyClosure(call=0x00000001166f93b8, op=0x0000000116b6e288, arglist=0x000000012038bc20, rho=0x00000001230468c8, suppliedvars=<unavailable>) at eval.c:1844:16 [opt]
    frame #67: 0x00000001051c7770 libR.dylib`bcEval(body=0x0000000116b7a528, rho=<unavailable>, useCache=<unavailable>) at eval.c:7104:12 [opt]
    frame #68: 0x00000001051dfb8c libR.dylib`R_compileAndExecute(call=0x00000001166f92a0, rho=0x00000001230468c8) at eval.c:1539:2 [opt]
    frame #69: 0x00000001051dee30 libR.dylib`do_for(call=0x00000001166f92a0, op=0x000000012300b988, args=0x00000001166f92d8, rho=0x00000001230468c8) at eval.c:2324:9 [opt]
    frame #70: 0x00000001051c0788 libR.dylib`Rf_eval(e=0x00000001166f92a0, rho=0x00000001230468c8) at eval.c:823:12 [opt]
    frame #71: 0x0000000105212d7c libR.dylib`Rf_ReplIteration(rho=0x00000001230468c8, savestack=<unavailable>, browselevel=<unavailable>, state=0x000000016b4d5ec0) at main.c:264:2 [opt]
    frame #72: 0x00000001052142ac libR.dylib`R_ReplConsole(rho=0x00000001230468c8, savestack=0, browselevel=0) at main.c:316:11 [opt]
    frame #73: 0x00000001052141e4 libR.dylib`run_Rmainloop at main.c:1139:5 [opt]
    frame #74: 0x0000000105214390 libR.dylib`Rf_mainloop at main.c:1146:5 [opt]
    frame #75: 0x000000010492bea0 R`main + 32
    frame #76: 0x0000000104c8d08c dyld`start + 520
```

**Comment by kevinushey** [comment]
Some other notes from macOS...

- The crash reproduces when using the CRAN build of R 4.2.1 + reticulate 1.25; both with versions compiled from source and the CRAN binary.
- The crash does not reproduce using Homebrew R (4.2.1) + reticulate 1.25.
- I also cannot reproduce with a locally-built version of R.

So it seems like (yikes) the crash could be somehow related to how R was configured / compiled?

I also never see this warning from rpy2 when using Homebrew R, or locally-compiled R:

> R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.

**Comment by kevinushey** [comment]
On Ubuntu 22.04, I see something different.

```
/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/rinterface_lib/embedded.py:260: UserWarning: R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
  warnings.warn(msg)
R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.
Warning in (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :
  libraries ‘/usr/local/lib/R/site-library’, ‘/usr/lib/R/site-library’ contain no packages
Hello!
Warning in (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :
  libraries ‘/usr/local/lib/R/site-library’, ‘/usr/lib/R/site-library’ contain no packages
Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: initialized

Current thread 0x0000ffff863c7020 (most recent call first):
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/rinterface.py", line 807 in __call__
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/rinterface_lib/conversion.py", line 45 in _
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/functions.py", line 30 in _formals_fixed
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/functions.py", line 133 in formals
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/functions.py", line 166 in __init__
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/__init__.py", line 224 in _rpy2py_sexpclosure
  File "/usr/lib/python3.10/functools.py", line 889 in wrapper
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/packages.py", line 264 in __fill_rpy2r__
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/packages.py", line 329 in __fill_rpy2r__
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/packages.py", line 203 in __init__
  File "/home/kevin/.virtualenvs/r-reticulate/lib/python3.10/site-packages/rpy2/robjects/packages.py", line 496 in importr
  File "<string>", line 1 in <module>

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, rpycall, _cffi_backend, markupsafe._speedups (total: 16)
Aborted (core dumped)
```

And with some extra frames on the stack:

```
gdb) bt
#0  __pthread_kill_implementation (threadid=281474842230816, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x0000fffff79df254 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x0000fffff799a67c in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x0000fffff7987130 in __GI_abort () at ./stdlib/abort.c:79
#4  0x0000ffffee815890 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#5  0x0000ffffee8158fc in _Py_FatalErrorFunc () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#6  0x0000ffffee7d56d8 in _Py_FatalError_TstateNULL () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#7  0x0000ffffee819950 in PyThreadState_Get () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#8  0x0000ffffee742e30 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#9  0x0000fffff7c92b84 in R_RunWeakRefFinalizer () from /usr/lib/R/lib/libR.so
#10 0x0000fffff7c92dd8 in ?? () from /usr/lib/R/lib/libR.so
#11 0x0000fffff7c56380 in Rf_eval () from /usr/lib/R/lib/libR.so
#12 0x0000fffff7c564d4 in Rf_eval () from /usr/lib/R/lib/libR.so
#13 0x0000fffff7beb050 in ?? () from /usr/lib/R/lib/libR.so
#14 0x0000fffff7bec880 in R_ToplevelExec () from /usr/lib/R/lib/libR.so
#15 0x0000fffff7bec8e8 in R_tryEval () from /usr/lib/R/lib/libR.so
#16 0x0000ffffd9661200 in _cffi_f_R_tryEval (self=<optimized out>, args=<optimized out>) at build/temp.linux-aarch64-cpython-310/_rinterface_cffi_api.c:3402
#17 0x0000ffffee7429cc in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#18 0x0000ffffee6fe840 in _PyObject_MakeTpCall () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#19 0x0000ffffee69922c in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#20 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#21 0x0000ffffee691748 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#22 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#23 0x0000ffffee700048 in _PyObject_FastCallDictTstate () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#24 0x0000ffffee70025c in _PyObject_Call_Prepend () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#25 0x0000ffffee768648 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#26 0x0000ffffee6fe840 in _PyObject_MakeTpCall () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#27 0x0000ffffee6981e8 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#28 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#29 0x0000ffffee696558 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#30 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#31 0x0000ffffee6965f0 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#32 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#33 0x0000ffffee700048 in _PyObject_FastCallDictTstate () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#34 0x0000ffffee70025c in _PyObject_Call_Prepend () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#35 0x0000ffffee7687e0 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#36 0x0000ffffee75d52c in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#37 0x0000ffffee6fe840 in _PyObject_MakeTpCall () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#38 0x0000ffffee6981e8 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#39 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#40 0x0000ffffee691748 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#41 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#42 0x0000ffffee6991e0 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#43 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#44 0x0000ffffee700848 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#45 0x0000ffffee69639c in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#46 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#47 0x0000ffffee700848 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#48 0x0000ffffee69639c in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#49 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#50 0x0000ffffee7000c0 in _PyObject_FastCallDictTstate () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#51 0x0000ffffee70025c in _PyObject_Call_Prepend () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#52 0x0000ffffee7687e0 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#53 0x0000ffffee75d52c in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#54 0x0000ffffee6fe840 in _PyObject_MakeTpCall () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#55 0x0000ffffee699abc in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#56 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#57 0x0000ffffee696558 in _PyEval_EvalFrameDefault () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#58 0x0000ffffee7d76d4 in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#59 0x0000ffffee7d2438 in PyEval_EvalCode () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#60 0x0000ffffee81d1ec in ?? () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#61 0x0000ffffee81e238 in PyRun_StringFlags () from /usr/lib/python3.10/config-3.10-aarch64-linux-gnu/libpython3.10.so
#62 0x0000fffff2c33174 in py_run_string_impl (code="from rpy2.robjects.packages import importr; importr('base').cat('Hello!\\n')", local=local@entry=false, convert=true) at python.cpp:2721
#63 0x0000fffff2c1ee0c in _reticulate_py_run_string_impl (codeSEXP=0xaaaab2315098, localSEXP=<optimized out>, convertSEXP=0xaaaab2b56d58) at RcppExports.cpp:544
#64 0x0000fffff7c04ac4 in ?? () from /usr/lib/R/lib/libR.so
#65 0x0000fffff7c44960 in ?? () from /usr/lib/R/lib/libR.so
#66 0x0000fffff7c56354 in Rf_eval () from /usr/lib/R/lib/libR.so
#67 0x0000fffff7c57ff0 in ?? () from /usr/lib/R/lib/libR.so
#68 0x0000fffff7c58e64 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#69 0x0000fffff7c4919c in ?? () from /usr/lib/R/lib/libR.so
#70 0x0000fffff7c56354 in Rf_eval () from /usr/lib/R/lib/libR.so
#71 0x0000fffff7c57ff0 in ?? () from /usr/lib/R/lib/libR.so
#72 0x0000fffff7c58e64 in Rf_applyClosure () from /usr/lib/R/lib/libR.so
#73 0x0000fffff7c4919c in ?? () from /usr/lib/R/lib/libR.so
#74 0x0000fffff7c5d3ec in ?? () from /usr/lib/R/lib/libR.so
#75 0x0000fffff7c5d9e0 in ?? () from /usr/lib/R/lib/libR.so
#76 0x0000fffff7c56678 in Rf_eval () from /usr/lib/R/lib/libR.so
#77 0x0000fffff7c87b88 in Rf_ReplIteration () from /usr/lib/R/lib/libR.so
#78 0x0000fffff7c88024 in ?? () from /usr/lib/R/lib/libR.so
#79 0x0000fffff7c880ec in run_Rmainloop () from /usr/lib/R/lib/libR.so
#80 0x0000aaaaaaaa0960 in main ()
#81 0x0000fffff79873fc in __libc_start_call_main (main=main@entry=0xaaaaaaaa0940 <main>, argc=argc@entry=3, argv=argv@entry=0xfffffffff138) at ../sysdeps/nptl/libc_start_call_main.h:58
#82 0x0000fffff79874cc in __libc_start_main_impl (main=0xaaaaaaaa0940 <main>, argc=3, argv=0xfffffffff138, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=<optimized out
```

**Comment by dfalbel** [comment]
I added a print statement like `std::cout << "Deleting object: " << (void*) pyObject << std::endl;` into:

https://github.com/rstudio/reticulate/blob/70a23d693736e7f82b1bbb484579f2849ce41ddd/src/reticulate_types.h#L12-L16

and it seems that the error is caused by a duplicated call that function for that same object.

```
Deleting object: 0x1434230e0
Deleting object: 0x1434230e0
```

It's hard to figure out why it's called twice. This is the deleter of the R External Pointer that holds the Python object. Calling GC will delete the external pointer and then register a finalizer that is then called at some later moment.

**Comment by lgautier** [comment]
I am less familiar with the `reticulate` code base. This is a guess more than anything.

From the above I understand that 2 R objects are external pointer addresses referencing the same Python object. Each time one of them is created the reference count for the Python should be increased by one. The only way I could see a race condition happen is if the GIL is released and incrementing/decrementing reference counts is allowed to happen in parallel (e.g. through threads).

Your call to R's `gc()` would force things to happen in sequence since R must be forcing all execution to happen in the main thread.

You could add to your debug prints a check of the reference count for the Python object (https://docs.python.org/3/c-api/structures.html#c.Py_REFCNT), and the GIL status (https://docs.python.org/3/c-api/init.html#c.PyGILState_Check).

**Comment by t-kalinowski** [comment]
@dfalbel and @lgautier, thank you for your help with this! The recent issue thread activity prompted me to start typing out my understanding of the issue, and while writing out three paragraphs hypothesizing that we're missing an `PyIncRef()` somewhere, I rubber-ducked myself into what is essentially a 1 line fix: a missing `GILscope` in the R external pointer finalizer. 🎉

**Comment by dfalbel** [comment]
Awesome! I was about to try that after reading @lgautier post. Glad it worked :)
